client/session/fssession/send.go

```diff
@@ -0,0 +1,211 @@
+	if sp.doneCh != nil {
+		doneCh = sp.doneCh
+		sp.doneCh = nil
+	}
+	err := pr.sendFn(stream, sp.root, excludes, progress)
+	if doneCh != nil {
+		if err != nil {
+			doneCh <- err
+		}
+		close(doneCh)
+	}
+	return err
+}
+
+func (sp *fsSendProvider) SetNextProgressCallback(f func(int, bool), doneCh chan error) {
+	sp.p = f
+	sp.doneCh = doneCh
+}
+
+type progressCb func(int, bool)
+
+type protocol struct {
+	name   string
+	sendFn func(stream grpc.Stream, srcDir string, excludes []string, progress progressCb) error
+	recvFn func(stream grpc.Stream, destDir string, cu CacheUpdater) error
+}
+
+func isProtoSupported(p string) bool {
+	if override := os.Getenv("BUILD_STREAM_PROTOCOL"); override != "" {
+		return strings.EqualFold(p, override)
+	}
+	return true
+}
+
+var supportedProtocols = []protocol{
+	{
+		name:   "tarstream",
+		sendFn: sendTarStream,
+		recvFn: recvTarStream,
+	},
+	{
+		name:   "diffcopy",
+		sendFn: sendDiffCopy,
+		recvFn: recvDiffCopy,
+	},
+}
+
+// FSSendRequestOpt defines options for FSSend request
+type FSSendRequestOpt struct {
+	SrcPaths         []string
+	OverrideExcludes bool
+	DestDir          string
+	CacheUpdater     CacheUpdater
+}
+
+// CacheUpdater is an object capable of sending notifications for the cache hash changes
+type CacheUpdater interface {
+	MarkSupported(bool)
+	HandleChange(fsutil.ChangeKind, string, os.FileInfo, error) error
+}
+
+// FSSend initializes a transfer of files
+func FSSend(ctx context.Context, name string, c session.Caller, opt FSSendRequestOpt) error {
+	var pr *protocol
+	for _, p := range supportedProtocols {
+		if isProtoSupported(p.name) && c.Supports(session.MethodURL(_FileSync_serviceDesc.ServiceName+"."+name, p.name)) {
+			pr = &p
+			break
+		}
+	}
+	if pr == nil {
+		return errors.Errorf("no fssend handlers for %s", name)
+	}
+
+	opts := make(map[string][]string)
+	if opt.OverrideExcludes {
+		opts["Override-Excludes"] = []string{"true"}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	client := newFileSyncClientWithName(name, c.Conn())
+
+	var stream grpc.ClientStream
+
+	ctx = metadata.NewContext(ctx, opts)
+
+	switch pr.name {
+	case "tarstream":
+		cc, err := client.TarStream(ctx)
+		if err != nil {
+			return err
+		}
+		stream = cc
+	case "diffcopy":
+		cc, err := client.DiffCopy(ctx)
+		if err != nil {
+			return err
+		}
+		stream = cc
+	}
+
+	return pr.recvFn(stream, opt.DestDir, opt.CacheUpdater)
+}
+
+// TODO: generate this, or use clientinterceptor
+type namedFileSyncClient struct {
+	cc   *grpc.ClientConn
+	name string
+}
+
+func newFileSyncClientWithName(name string, cc *grpc.ClientConn) FileSyncClient {
+	return &namedFileSyncClient{cc, name}
+}
+
+func (c *namedFileSyncClient) DiffCopy(ctx context.Context, opts ...grpc.CallOption) (FileSync_DiffCopyClient, error) {
+	stream, err := grpc.NewClientStream(ctx, &_FileSync_serviceDesc.Streams[0], c.cc, "/"+_FileSync_serviceDesc.ServiceName+"."+c.name+"/DiffCopy", opts...)
```

Comment:
@stevvooe What we are doing is dynamically setting the service name. It is still `/<package>.<Service>/<Method>` (wire definition does not define package part) just the service name is updated. If you generate another client you just need to specify the full name(in your proto for example) and everything will work properly.

> Why don't we define a service for each protocol?

This is not about adding protocols. Every protocol is in own method, if there will ever be more we will just define more methods. It is about reusing the same service definitions for multiple(not currently implemented) file sync sources. Similar example from normal grpc usage would be if you have a generic CRUD service definition with `Create()/Read()` methods and then you want to expose different object services without redefining the service definition. Normally this is done with copy&paste but in this case because the initialization is based on configuration from client, we don't know how many services there will be exposed.

I think I could do the file syncing with just `NotImplemented` with a bit more complicated code for retries and extra roundtrip.The authorization feature seems more complicated as that requires completely different code path depending if token part is supported or not. I guess we would need to refine extra `IsXSupported()` method then for that service. We would not be able to document what can be done using this connection as everything is allowed with the caveat that one side can't expect the other side to perform anything correctly. I could also do it with health check, not sure if you are ok with reusing the default server implementation with a string that contains the instance/method name though. If I'm thinking about someone possibly reimplementing a client, doing the healthcheck, extra call or retries seems considerably more difficult.